cleangoldenami/main.go

```diff
@@ -96,40 +112,74 @@ func main() {
 							x86Images = append(x86Images[:imageToDeleteIndex], x86Images[imageToDeleteIndex+1:]...)
 						}
 					}
-					if !*dryRun {
-						fmt.Printf("Region %v image capacity is currently at %v out of %v:\n", regionName, len(images), quota)
-					} else {
-						fmt.Printf("Region %v image capacity is currently at %v out of %v - would delete the following AMIs:\n", regionName, len(images), quota)
+					drr := deregisterRegionResult{
+						region:                 regionName,
+						currentImages:          len(images),
+						quota:                  quota,
+						deregisterImageResults: make([]deregisterImageResult, len(imagesToDelete)),
 					}
-					for _, image := range imagesToDelete {
-						for _, t := range image.Tags {
-							if *t.Key == "version" {
-								if !*dryRun {
-									err = deregisterImage(ec2Client, image)
-									if err != nil {
-										fmt.Printf("  - error deregistering image %v (%v): %v\n", *image.ImageId, *t.Value, err)
-									} else {
-										fmt.Printf("  - successfully deregistered image %v (%v)\n", *image.ImageId, *t.Value)
+
+					deregisterWg := sync.WaitGroup{}
+					for index, image := range imagesToDelete {
+						wg.Add(1)
+						go func(img ec2Types.Image, imageIndex int) {
+							defer wg.Done()
+							for _, t := range img.Tags {
+								if *t.Key == "version" {
+									var deregisterImageErr error
+									if !(*dryRun) {
+										deregisterImageErr = deregisterImage(ec2Client, img)
 									}
-								} else {
-									fmt.Printf("  - %v (%v)\n", *image.ImageId, *t.Value)
+									drr.deregisterImageResults[imageIndex] = deregisterImageResult{imageId: *img.ImageId, version: *t.Value, err: deregisterImageErr}
+									break
 								}
-								break
 							}
-						}
+						}(image, index)
 					}
+					deregisterWg.Wait()
+
+					deregisterRegionResultChan <- drr
 				} else {
-					fmt.Printf("ERROR: MANUAL ACTION REQUIRED - Unable to delete images in region %v (Total Images: %v, rhel-arm64: %v, legacy-x86_64: %v, rhel-x86_64: %v, Quota: %v, ImagesToDelete: %v)\n",
+					fmt.Printf("ERROR: MANUAL ACTION REQUIRED - Unable to delete images in region %v (Total Images: %v, rhel-arm64: %v, legacy-x86_64: %v, rhel-x86_64: %v, Quota: %v, NumImagesToDelete: %v)\n",
 						regionName, imageCount, len(arm64Images), len(legacyx86Images), len(x86Images), quota, numOfImagesToDelete)
 				}
 			}
 		}(*enabledRegion.RegionName)
 	}
 
 	wg.Wait()
+
+	var deregisterResults []deregisterRegionResult
```

Comment:
You know how long this needs to be, so you can:

```golang
deregisterResults := make([]deregisterRegionResult, 0, len(enabledRegions))
```

instead of growing it on the fly as you append.  Maybe the Go compiler is smart enough not to care?  But if you're explicit, you don't need to worry about how smart the Go compiler is at the moment ;)

Reply:
Great suggestion - and I _would_ do this, but I just rewrote this block to not need the slice at all 😅